### PR TITLE
Bug 1887745: Listing All Events with Chunking

### PR DESCRIPTION
**Pull request by otaviof**
Using chunking/pagination to `LIST` all events from the API Server.

### CODE REVIEW

File: pkg/unidling/controller/unidling_controller.go
Comment (gabemontero) [inline pkg/unidling/controller/unidling_controller.go:102]: were you able to confirm that you still got add events with this @otaviof ? 

and is it that with this change, do you no longer got the relists on the the `resyncPeriod` interval ?

Comment (otaviof) [reply pkg/unidling/controller/unidling_controller.go:102]: @gabemontero yes, positve. It does handle new events. And also, it does not issue `LIST` on `resyncPeriod` and during client bootstrap.

Comment (gabemontero) [reply pkg/unidling/controller/unidling_controller.go:102]: OK good.  Then I suggest updating that comment to note a) we only care about add events, and b) this form of informer initialization will still get us add events, but avoid lists on the resync interval etc.

Comment (otaviof) [reply pkg/unidling/controller/unidling_controller.go:102]: Sure, thanks for the suggestion. Please consider updated description.

Comment (coreydaley) [reply pkg/unidling/controller/unidling_controller.go:102]: I would put the full url to the bugzilla, "Bug <number>" seems to ambiguous.

Comment (otaviof) [reply pkg/unidling/controller/unidling_controller.go:103]: @coreydaley okay, I'm adding the full Bugzilla URL in the comment. Please consider.

Comment (gabemontero) [reply pkg/unidling/controller/unidling_controller.go:103]: can you add the details I asked for in https://github.com/openshift/openshift-controller-manager/pull/145/files#r516818486 @otaviof 

`a) we only care about add events, and b) this form of informer initialization will still get us add events, but avoid lists on the resync interval etc.`

Comment (otaviof) [reply pkg/unidling/controller/unidling_controller.go:103]: @gabemontero It was added to the PR description, but now also as a comment as well. Duly noted.

Comment (gabemontero) [reply pkg/unidling/controller/unidling_controller.go:103]: /lgtm

/assign @adambkaplan 
(though he may be out for a bit ... but I'll assign him in case he sees this and the discussion above prompts him to approve)
/assign @deads2k 
(as `git blame` cites him *from 5 years ago* for the prior line 102 `// No need to list -- we only care about new events` ... ideally we get a sanity check from him that our interpretation of things is correct)

Comment (otaviof) [reply pkg/unidling/controller/unidling_controller.go:103]: Thank you, folks!

**Comment by adambkaplan** [comment]
/hold

Without a re-list, can't the unidling controller miss a "NeedPods" event that it ultimately is looking for?

That said, watching events itself feels like an anti-pattern for a controller. We've come a long way in the past 5 years in terms of understanding and best practices...

**Comment by adambkaplan** [comment]
> That said, watching events itself feels like an anti-pattern for a controller. We've come a long way in the past 5 years in terms of understanding and best practices...

Note that getting rid of this controller or completely re-working its mechanics is beyond the scope of this bug fix.

Bump @mfojtik @deads2k

File: pkg/unidling/controller/unidling_controller.go
Comment (sttts) [reply pkg/unidling/controller/unidling_controller.go:103]: How does this watch without a ListWatch?

File: pkg/unidling/controller/unidling_controller.go
Comment (otaviof) [reply pkg/unidling/controller/unidling_controller.go:105]: Please consider the latest changes, we are only loading 500 events at once now.

Comment (sttts) [reply pkg/unidling/controller/unidling_controller.go:105]: I don't think you have a guarantee that these events are the newest ones. We still have to get all of them, but with pagination. See slack thread where I shared links showing how that works.

**Comment by otaviof** [comment]
@adambkaplan we had a successful CI run ([unit](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_openshift-controller-manager/145/pull-ci-openshift-openshift-controller-manager-master-unit/1328705234632970240), [e2e-aws](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_openshift-controller-manager/145/pull-ci-openshift-openshift-controller-manager-master-e2e-aws/1328705234507141120) and [e2e-aws-upgrade](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_openshift-controller-manager/145/pull-ci-openshift-openshift-controller-manager-master-e2e-aws-upgrade/1328705234540695552)) using only 3 events per request. I'm moving it back to default limit.

File: pkg/unidling/controller/unidling_controller.go
Comment (otaviof) [reply pkg/unidling/controller/unidling_controller.go]: @sttts in fact it would not properly watch without a `WatchFunc` in the `cache.ListWatch`. The attempt before was removing too many elements from this Informer.

File: pkg/unidling/controller/unidling_controller.go
Comment (otaviof) [reply pkg/unidling/controller/unidling_controller.go:105]: @sttts thanks for the links on Slack. Now I can understand how this Informer is wired up and based on the examples in the Slack thread, and with @adambkaplan help, we are chunking the listing requests. Please consider latest changes.

**Comment by otaviof** [comment]
@sttts, @adambkaplan Would you please consider latest commit using [client-go Pager](https://pkg.go.dev/k8s.io/client-go@v0.19.0-rc.3/tools/pager)? It implements the chunking logic on [`List` function](https://github.com/otaviof/openshift-controller-manager/blob/6a7ee49e451bae4c444ba3def0a53da6644a0cea/vendor/k8s.io/client-go/tools/pager/pager.go#L73-L145).

File: pkg/unidling/controller/unidling_controller.go
Comment (otaviof) [inline pkg/unidling/controller/unidling_controller.go:25]: @isutton thanks for showing the client-go Pager implementation, appreciated!

File: pkg/unidling/controller/unidling_controller.go
Comment (sttts) [reply pkg/unidling/controller/unidling_controller.go:105]: This looks good! :+1:

File: pkg/unidling/controller/unidling_controller.go
Comment (adambkaplan) [inline pkg/unidling/controller/unidling_controller.go:113]: Perhaps mark a TODO comment - we should wire a global context through to the controller.

File: pkg/unidling/controller/unidling_controller.go
Comment (sttts) [inline pkg/unidling/controller/unidling_controller.go:111]: "to increase"? what does it mean?

Comment (otaviof) [reply pkg/unidling/controller/unidling_controller.go:111]: @sttts means a temporary change to make sure it will trigger "chunking". I'll clean this up.

Comment (otaviof) [reply pkg/unidling/controller/unidling_controller.go]: Just pushed the changes in a single commit.

File: pkg/unidling/controller/unidling_controller.go
Comment (otaviof) [reply pkg/unidling/controller/unidling_controller.go:115]: Sure. I'm adding a comment to highlight.

**Comment by adambkaplan** [comment]
/hold

Maintaining the hold because it appears that there's an upstream bug that is preventing this PR from working correctly (events endpoint is not respecting `resourceVersion=0`, and is returning all events instead of the chunked list). Once that's been filed IMO we can merge this PR (it does zero harm).

**Review by adambkaplan — APPROVED** [review]
/lgtm

**Comment by otaviof** [comment]
> Maintaining the hold because it appears that there's an upstream bug that is preventing this PR from working correctly (events endpoint is not respecting `resourceVersion=0`, and is returning all events instead of the chunked list). Once that's been filed IMO we can merge this PR (it does zero harm).

To clarify, the events endpoint is respecting the [Storage Cache rules](https://github.com/kubernetes/apiserver/blob/0e46f0ea2bdc958827712205cf9a88140a28409d/pkg/storage/cacher/cacher.go#L655-L668), as we can read on the comment (pointed out by @sttts). Sorry for the misinformation early on.

So, the purposed approach is to replace `resourceVersion=0` with empty string, to make sure the controller will employ chunking on bootstrap and resync periods, at the expense of avoiding the storage cache.

File: pkg/unidling/controller/unidling_controller.go
Comment (sttts) [inline pkg/unidling/controller/unidling_controller.go:116]: would mention that `=0` means to list from cache and the cache does not support chunking.

Comment (otaviof) [reply pkg/unidling/controller/unidling_controller.go]: Indeed. Just amended the comment to mention zero means reading from cache. Thanks!

File: pkg/unidling/controller/unidling_controller.go
Comment (sttts) [inline pkg/unidling/controller/unidling_controller.go:117]: nit: ... from cache, which does not support chunking

**Review by adambkaplan — APPROVED** [review]
/hold cancel

No need to file an upstream issue, as we uncovered the special case of `ResourceVersion=0`.

**Pull request merged**
